### PR TITLE
Link email with complaint using Amazon SES webhook

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
Currently there is no information about which email generated a complaint using the AWS API transport. This PR adds the ability to associate an email with a complaint.

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Send an email to contact using any email service
3. Switch email service configuration to Amazon SES - API
4. Simulate SES notification with request (replace recipient email and X-EMAIL_ID):
```
POST https://mautic.ddev.site/mailer/amazon_api/callback
{
  "notificationType": "Complaint",
  "complaint": {
    "feedbackId": "010276757657bd3d-31765657-e767-4722-9957-b765765654d7-000000",
    "complaintSubType": null,
    "complainedRecipients": [
      {
        "emailAddress": "recipien@example.com"
      }
    ],
    "timestamp": "2023-02-02T21:37:43.000Z"
  },
  "mail": {
    "timestamp": "2023-02-02T09:40:21.755Z",
    "source": "\"Example\" <sender@example.com>",
    "messageId": "01027657675f9e7b-1c76575f-1659-4651-b650-87e3f76565e1-000000",
    "destination": [
      "recipien@example.com"
    ],
    "headersTruncated": false,
    "headers": [
      {
        "name": "X-EMAIL-ID",
        "value": "18"
      }
    ]
  }
}
```
5. Check contacts details view, complaint should be linked with email
image

### CODE REVIEW

**Comment by mabumusa1** [comment]
@patrykgruszka your code changes are correct, I have the complete code for SES on this repo.

These code changes will be merged on M5.

https://GitHub.com/mabumusa1/ses-plugin/

**Comment by patrykgruszka** [comment]
@mabumusa1 I see you already check the X-EMAIL-ID in your plugin. Does this work without checking the type from the `notificationType` key? I'm not sure, but the AWS docs (https://docs.aws.amazon.com/ses/latest/dg/notification-examples.html) say that the type should be in the `notificationType` key. Anyway, it would be worth including the additional unit tests I've written.

**Comment by mabumusa1** [comment]
You always need to check the notification type, based on it, you decide how to handle the complaint or the bounce


Also it is needed if there is a configset involved.

I have been using the plugin in production for the past five months, and I think it covers all the test cases.

File: app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php [hunk 1]
```diff
@@ -51,12 +51,16 @@ public function processCallbackRequest(Request $request)
...
-            throw new HttpException(400, "Key 'Type' not found in payload ");
-        }
-
         // determine correct key for message type (global or via ConfigurationSet)
-        $type = (array_key_exists('Type', $payload) ? $payload['Type'] : $payload['eventType']);
+        if (isset($payload['Type'])) {
+            $type = $payload['Type'];
+        } elseif (isset($payload['eventType'])) {
+            $type = $payload['eventType'];
+        } elseif (isset($payload['notificationType'])) {
```
Comment (patrykgruszka) [inline app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:59]: @mabumusa1 I'm talking about this change. Before this PR and in your plugin, only the `Type` and `eventType` headers are checked. Are complaints correctly recognised in your production environment without this change?

Comment (mabumusa1) [reply app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:59]: @patrykgruszka it is first checked to verify that the payload is correct, then I re-check the content of the payload depending on its type 

Check this function https://github.com/mabumusa1/ses-plugin/blob/main/Mailer/Callback/AmazonCallback.php#L85

`notificationType` is not set at the same level as `Type` and `eventType` this is why I check them first, then I process the payload depending on the existence of `Type` or `eventType`

Comment (patrykgruszka) [reply app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:59]: @mabumusa1 You're right, so if you've already implemented this feature, I think we can close this PR, right?

Comment (mabumusa1) [reply app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:59]: @patrykgruszka we can close this one, please feel free to use the code changes on the plugin. I will release it with M5

**Pull request closed without merging**